src/content/blog/post-mortem-blob-storage-outage.md

```diff
@@ -0,0 +1,40 @@
+---
+title: "Post-mortem: Blob Storage Outage"
+generated: 1701894028874
+description: Disclosure, mitigation, and next steps
+pubDate: May 2, 2024
+author: Steve Krouse
+---
+
+Val Town [provides blob storage](https://docs.val.town/std/blob/) for use within Vals. Blob storage is used by many Vals as a quick key value store or as a place to store larger files like images and logs. Here's a basic example:
+
+```tsx
+import { blob } from "https://esm.town/v/std/blob";
+
+await blob.setJSON("myKey", { isBlob: true });
+const { isBlob } = await blob.getJSON("myKey");
+console.log(isBlob)
+```
+
+Under the hood, blob storage is backed by [Cloudflare's S3-compatible object storage offering: R2](https://www.cloudflare.com/developer-platform/r2/).
+
+On April 30th at 5:05pm UTC while offboarding a Val Town employee, we removed their user account from our Cloudflare organization. The API credentials that we had been using to communicate with R2 were tied to the user account, and stopped working once the account was removed. At 8:00pm UTC we had received a handful of user-reports about unexpected blob storage errors. Over the next hour, we raised an alert internally, identified the issue and replaced the credentials.
+
+### Impact
+
+Blob storage was unavailable from 5:05pm UTC until 9:05pm UTC. During that time 3,492 read requests and and 135 write requests failed to communicate with blob storage.
+
+### Cause
+
+Cloudflare's API credentials are tied to user accounts. When clicking through Cloudflare's R2 dashboard there is an API Tokens section where you can issue S3-compatible API credentials. It is not entirely clear that this page is scoped to an individual user and not the Val Town account itself. We were unaware of the scope of the credentials and did not appropriately ensure that the API credentials were tied to a long-lasting account.
+
+[Cloudflare recommends](https://community.cloudflare.com/t/how-to-create-an-api-token-not-linked-to-a-user/556088/4) that organizations create a system account for credential creation.
```

Comment:
This seems like such a huge footgun, right? Are companies making our mistake all the time? It'd be interesting if we could find other such postmortems – otherwise I feel like this post could serve partly as a PSA for folks about this footgun